specs/language/declarations.tex

```diff
@@ -0,0 +1,44 @@
+\Ch{Declarations}{Decl}
+\Sec{Function Definitions}{Decl.Function}
+\Sec{Attributes}{Decl.Attr}
+\Sub{Entry Attributes}{Decl.Attr.Entry}
+\Sec{Linkage}{Decl.Linkage}
+
+\p An entity that denotes an object, reference, function, type, template, namespace, or value, may have a \textit{linkage}. If a name has \textit{linkage}, it refers to the same entity as the same name introduced by a declaration in another scope. If a variable, function, or another entity with the same name is declared in several scopes, but does not have sufficient \textit{linkage}, then several instances of the entity are generated.
+
+\p There are three linkages recognized: \textit{external linkage}, \textit{internal linkage} and \textit{no linkage}.
+
+\Sub{External Linkage}{Decl.Linkage.External}
+
+\p Entities with \textit{external linkage} can be referred to from the scopes in the other translation units and enable linking between them.
+
+\p The following entities in HLSL have \textit{external linkage}:
+\begin{itemize}
+    \item entry point functions
+    \item functions marked with \texttt{export} keyword
+    \item global variables that are not marked \texttt{static} or \texttt{groupshared} \footnote{These are not really linked with other translation units but rather their values are loaded indirectly based on cbuffer mapping.}
+    \item enumerations
+    \item classes or template classes, their member functions, static data members, nested classes and enumerations
+\end{itemize}
+
+\p Linkage of functions (including template functions) that are not entry points or marked with \texttt{export} keyword is implementation dependent. \footnote{In DXC today functions that are not entry points or exported have \textit{internal linkage} by default. This can be overriden by \texttt{-default-linkage} compiler option.}
+    
+\Sub{Internal Linkage}{Decl.Linkage.Internal}
+
+\p Entities with \textit{internal linkage} can be referred to from all scopes in the current translation unit.
+
+\p The following entities in HLSL have \textit{internal linkage}:
+\begin{itemize}
+    \item global variables marked as \texttt{static} or \texttt{groupshared}
```

Comment:
Unnamed namespaces and members of unnamed namespaces also are given internal linkage.

Reply:
DXC does not support unnamed namespaces. Should we support them in Clang?

Reply:
That's an interesting question... It hadn't occurred to me that DXC didn't support them. They work in Clang today because we haven't done anything to actively disable them. My inclination is that unless there is a reason not to support them, we should just plan them as a 202x or 202y feature. I would lean toward 202x, just because I think it is trivial to enable this in DXC:

```diff
diff --git a/tools/clang/lib/Parse/ParseDeclCXX.cpp b/tools/clang/lib/Parse/ParseDeclCXX.cpp
index de46dfdb43c..be072a0715d 100644
--- a/tools/clang/lib/Parse/ParseDeclCXX.cpp
+++ b/tools/clang/lib/Parse/ParseDeclCXX.cpp
@@ -94,7 +94,8 @@ Decl *Parser::ParseNamespace(unsigned Context,
     }
   }
   // HLSL Change Starts
-  else if (getLangOpts().HLSL) {
+  else if (getLangOpts().HLSL &&
+           getLangOpts().HLSLVersion < hlsl::LangStd::v202x) {
     Diag(Tok, diag::err_expected) << tok::identifier;
   }
   // HLSL Change Ends
```

Curious if @Tex3d has any historical context for why anonymous namespaces were disabled other than matching FXC.

Reply:
I've updated the spec to mention that entities in unnamed namespaces have internal linkage.

Reply:
I believe the original context was just to match FXC support for HLSL.  FXC can't parse an anonymous namespace, so we disabled it originally in DXC.  We never prioritized re-enabling it, so it's still that way.